### ISSUE TITLE
Feature Request: oEmbed API

### ISSUE THREAD

**Issue by Ipstenu** [netPhotoGraphics/netPhotoGraphics#1341]
Sometimes I want to embed a photo on another webpage like one might embed a tweet.

Right now I would:

1. Download the thumbnails
2. Upload them to whatever blog
3. Edit the photo to link to my gallery

It would be nice if I could paste: example.com/gallery/album/image.html

Adding oEmbed to WordPress is pretty trivial for me (and I'm happy to run that end and even make a pull request to WP core for it if we get it rolling). It would allow for greater integration with external sites/blogs.

I do know how monumental that work is, to get the API up, so I know this is not a trivial thing!

**Comment by sbillard** [comment]
I am not a Wordpress user, so I looked up oEmbed. I may be missing something, but the results I found seem to indicate that all the heavy lifting for this is on the Wordpress size. oEmbed has to "support" the service (whatever that means.)

Anyway, if you are willing to educate me, we can probably make this work. netPhotoGraphics has a plugin named _externamFeed_ that might be a help. It can retrieve individual sized images. It is kind of a programmatic RSS. I don't think the plugin has been used recently, so it could have grown issues.

**Comment by Ipstenu** [comment]
The 'support' part (which yeah, they really could do a better job explaining that one!) is the part that has to be in NPG (or as a plugin). 

The most digestible explanation I found was https://www.ycombinator.com/library/3v-how-to-build-an-oembed-integration-for-your-startup


1. NPG builds a special version of a post (album or image) that is designed to live in an iframe.
2. NPG has a 'get' endpoint that knows to take parameters (url, maxwidth, maxheight, and format) and kicks out an JSON response
3. That JSON response includes the HTML of the iframe (based on item 1)

For example, if you looked on my WordPress site here's a recent post:

https://halfelf.org/2021/hugo-deployment-via-github-actions/

That includes this:

```
   <link rel="alternate" type="application/json+oembed" href="https://halfelf.org/wp-json/oembed/1.0/embed?url=https%3A%2F%2Fhalfelf.org%2F2021%2Fhugo-deployment-via-github-actions%2F"/>
    <link rel="alternate" type="text/xml+oembed" href="https://halfelf.org/wp-json/oembed/1.0/embed?url=https%3A%2F%2Fhalfelf.org%2F2021%2Fhugo-deployment-via-github-actions%2F&#038;format=xml"/>
```

Which is so the oembed is 'discoverable' by other apps.

And that URL is what kicks out the 'embed' content:

https://halfelf.org/wp-json/oembed/1.0/embed?url=https%3A%2F%2Fhalfelf.org%2F2021%2Fhugo-deployment-via-github-actions%2F

```
{
  "version": "1.0",
  "provider_name": "Half-Elf on Tech",
  "provider_url": "https://halfelf.org",
  "author_name": "Ipstenu (Mika Epstein)",
  "author_url": "https://halfelf.org/author/ipstenu/",
  "title": "Hugo Deployment via GitHub Actions - Half-Elf on Tech",
  "type": "rich",
  "width": 600,
  "height": 338,
  "html": "<blockquote class=\"wp-embedded-content\"><a href=\"https://halfelf.org/2021/hugo-deployment-via-github-actions/\">Hugo Deployment via GitHub Actions</a></blockquote>\n<script type='text/javascript'>\n<!--//--><![CDATA[//><!--\n\t\t/*! This file is auto-generated */\n\t\t!function(c,d){\"use strict\";var e=!1,n=!1;if(d.querySelector)if(c.addEventListener)e=!0;if(c.wp=c.wp||{},!c.wp.receiveEmbedMessage)if(c.wp.receiveEmbedMessage=function(e){var t=e.data;if(t)if(t.secret||t.message||t.value)if(!/[^a-zA-Z0-9]/.test(t.secret)){for(var r,a,i,s=d.querySelectorAll('iframe[data-secret=\"'+t.secret+'\"]'),n=d.querySelectorAll('blockquote[data-secret=\"'+t.secret+'\"]'),o=0;o<n.length;o++)n[o].style.display=\"none\";for(o=0;o<s.length;o++)if(r=s[o],e.source===r.contentWindow){if(r.removeAttribute(\"style\"),\"height\"===t.message){if(1e3<(i=parseInt(t.value,10)))i=1e3;else if(~~i<200)i=200;r.height=i}if(\"link\"===t.message)if(a=d.createElement(\"a\"),i=d.createElement(\"a\"),a.href=r.getAttribute(\"src\"),i.href=t.value,i.host===a.host)if(d.activeElement===r)c.top.location.href=t.value}}},e)c.addEventListener(\"message\",c.wp.receiveEmbedMessage,!1),d.addEventListener(\"DOMContentLoaded\",t,!1),c.addEventListener(\"load\",t,!1);function t(){if(!n){n=!0;for(var e,t,r=-1!==navigator.appVersion.indexOf(\"MSIE 10\"),a=!!navigator.userAgent.match(/Trident.*rv:11\\./),i=d.querySelectorAll(\"iframe.wp-embedded-content\"),s=0;s<i.length;s++){if(!(e=i[s]).getAttribute(\"data-secret\"))t=Math.random().toString(36).substr(2,10),e.src+=\"#?secret=\"+t,e.setAttribute(\"data-secret\",t);if(r||a)(t=e.cloneNode(!0)).removeAttribute(\"security\"),e.parentNode.replaceChild(t,e)}}}}(window,document);\n//--><!]]>\n</script><iframe sandbox=\"allow-scripts\" security=\"restricted\" src=\"https://halfelf.org/2021/hugo-deployment-via-github-actions/embed/\" width=\"600\" height=\"338\" title=\"&#8220;Hugo Deployment via GitHub Actions&#8221; &#8212; Half-Elf on Tech\" frameborder=\"0\" marginwidth=\"0\" marginheight=\"0\" scrolling=\"no\" class=\"wp-embedded-content\"></iframe>",
  "thumbnail_url": "https://i1.wp.com/halfelf.org/wp-content/uploads/sites/2/2021/09/david-gavi-EEuhA0VpGDk-unsplash.jpg?fit=1200%2C675&ssl=1",
  "thumbnail_width": 1200,
  "thumbnail_height": 675,
  "description": "Using GitHub actions to build and deploy a static site is fun."
}
```

That hefty chunk in "html" is the magic sauce. WordPress has a blockquote 'fallback' but the iframe is:

```
<iframe sandbox=\"allow-scripts\" security=\"restricted\" src=\"https://halfelf.org/2021/hugo-deployment-via-github-actions/embed/\" width=\"600\" height=\"338\" title=\"&#8220;Hugo Deployment via GitHub Actions&#8221; &#8212; Half-Elf on Tech\" frameborder=\"0\" marginwidth=\"0\" marginheight=\"0\" scrolling=\"no\" class=\"wp-embedded-content\"></iframe>
```

Which you can see is calling `https://halfelf.org/2021/hugo-deployment-via-github-actions/embed/` and if you look at that URL you see a heavily truncated version of the site.

So you end up needing two 'new' urls:

1. `https://example.com/gallery/album/image.html?embed` -- the iframe version of it all
2. `https://example.com/gallery/api/embed?url=...` - the json returner that gets added as a link= call

 I _think_ the gotchas would be:

1. Designing the default output (as you can see, WP's is messy)
2. Deciding if that should be customizable by a theme (I'd say yes, but that makes more work)
3. Oh and making the API URL because there really isn't one (?) in NPG

The original plugin that built WP's code is here: https://github.com/swissspidy/oEmbed-API

And there's a Zenphoto plugin that kind of does this: https://github.com/deanmoses/zenphoto-json-rest-api

I haven't looked too much into that plugin but that's my weekend to-do and maybe I can get a working prototype that is more 2021.

As for teaching WordPress what to do, it's a matter of telling WP "If someone pastes in a URL, check for the oembed discovery." Which it already does. And it knows "OH! I got a return AND it has HTML? Good! I'll echo that!" But that's all on WP's end, and I'm pretty sure it would be able to do it automagically since discovery exists by default there (and if not? Well that's 100% on me to solve!)

**Comment by sbillard** [comment]
The json plugin looks like a promising start. Another resource to look at is the externalFeed plugin. It is not JSON, though. More of an RSS type API. (I noticed some typos in the externalFeed document block--those are fixed in the repository, but I have not make a release including the changes.)

I would be interested in contributing to this effort, but probably you should take the lead. For instance, I presume the JSON plugin will need to be migrated to nPG.

**Comment by Ipstenu** [comment]
Yep made a plugin

https://github.com/JorjaFox/embed-npg

It works, more or less, but it's VERY basic.

**Comment by sbillard** [comment]
I made a pull request to your repository. Some things that the legacy converter did not properly handle.

**Comment by Ipstenu** [comment]
I think the only 'TODO' that my brain doesn't want to wrap about would be making a proper 'endpoint.'

Right now, I've got it using `example.com/albumb/image.html?json-oembed`

But really it would be better as `example.com/oembed/?url=https://example.com/albumb/image.html`

Is there a way to do that?

**Comment by sbillard** [comment]
The plugin could create a rewrite rule that would handle this. I will have to look at how your plugin works to see what the rewrite should be. I should have time tomorrow.

**Comment by sbillard** [comment]
How about the link being: `example.com/oembed/album/image.html`? I  can make rewrite rules that turn that into `example.com/album/image.html`. Same for _json-oembed_.

In testing this I note you have reverence to _themes/basic/images/oembed-icon.png_. That should not be hardcoded to the theme. Also there should be a default icon. I will make a pull request shortly with these updates.

**Comment by Ipstenu** [comment]
I was in the middle of making a better format for embedded gallery with images (vs one with sub albums) so that was a moment of jiggery to fold it all in together. 

I tweaked it slightly to allow overriding by your theme (if desired, obv)

https://github.com/JorjaFox/embed-npg/blob/production/oembed.php#L460-L466

And then I ended up having to wrap `ob_end_clean ` like this:
```
		if (ob_get_length() > 0 ) {
			ob_end_clean();
		}
```

If I didn't, it kept throwing alerts in my debug log, and that appears to fix it.

**Comment by sbillard** [comment]
My bad. ob_get_clean() ends the buffering. So, the whole code you changed can just be removed. Not sure why I did not get any errors in my tests, though.

**Comment by sbillard** [comment]
I've fixed this and also implemented customization for the iFrame and oEmbed for the gallery page. Pull request is pending.

**Comment by Ipstenu** [comment]
Pulled in. This, this is awesome. 

I figured out how to do options so now you can adjust the iframe height/width, with decent defaults.

In order to make it not overflow boundaries, though, I had to remove the break to make the row of images on a new line. With it, it would over-flow hid on WordPress and Drupal (which it's supposed to). But to make the images fit properly, I made them smaller. Since they now have the album thumbnail, it works out, and it let me bump the images to 5.

That will still cause small issues when there are images and albums (see https://jorjafox.net/gallery/embed/tv/csi-vegas/publicity/01/ for a live example)

I think I'll put in an option to turn it on, since I can see some people wanting to show it all, and adjust things, while others won't.

**Comment by Ipstenu** [comment]
Yeah, did that. I also ran into a weird bug with the rewrite.

Since I run my gallery in a folder, the mod_rewrite code would say that my oembed URL was `example.com/embed/GALLERY/album/`  instead of `example.com/GALLERY/embed/album/`

This is how I fixed it, but I don't know if that's really shitty form. I wasn't able to figure out how to override otherwise.

```
		if (MOD_REWRITE) {
			if (WEBPATH) {
				$link = str_replace( WEBPATH, '', $link );
			}
			$path = WEBPATH . '/' . $rewrite . $link;
		} else {
```
